pkg/volume/glusterfs/glusterfs.go

```diff
@@ -528,20 +563,113 @@ func (p *glusterfsVolumeProvisioner) CreateVolume() (r *api.GlusterfsVolumeSourc
+	for _, node := range clusterinfo.Nodes {
+		nodei, err := cli.NodeInfo(string(node))
+		if err != nil {
+			glog.Errorf("glusterfs: failed to get hostip %s ", err)
+			return nil, 0, fmt.Errorf("failed to get hostip %v", err)
+		}
+		ipaddr := dstrings.Join(nodei.NodeAddRequest.Hostnames.Storage, "")
+		dynamicHostIps = append(dynamicHostIps, ipaddr)
+	}
+	glog.V(1).Infof("glusterfs: hostlist :%v", dynamicHostIps)
+	if len(dynamicHostIps) == 0 {
+		glog.Errorf("glusterfs: no endpoint hosts found %s ", err)
+		return nil, 0, fmt.Errorf("no endpoint hosts found %v", err)
+	}
+
+	// The 'endpointname' is created in form of 'cluster-<id>' where 'id' is the cluster id.
+	// CreateEndpointService() checks for this 'endpoint' existence in PVC's namespace and
+	// If not found, it create an endpoint and svc using the IPs we dynamically picked at time
+	// of volume creation.
+	epServiceName := dynamicEpSvcPrefix + volume.Cluster
+	epNamespace := p.options.PVC.Namespace
+	endpoint, service, err := p.CreateEndpointService(epNamespace, epServiceName, dynamicHostIps)
+	if err != nil {
+		glog.Errorf("glusterfs: failed to create endpoint/service")
+		return nil, 0, fmt.Errorf("failed to create endpoint/service %v", err)
+	}
+	glog.V(1).Infof("glusterfs: dynamic ep %#v and svc : %#v ", endpoint, service)
+
 	return &api.GlusterfsVolumeSource{
-		EndpointsName: p.endpoint,
+		EndpointsName: endpoint.Name,
 		Path:          volume.Name,
 		ReadOnly:      false,
 	}, sz, nil
 }
 
+func (p *glusterfsVolumeProvisioner) CreateEndpointService(namespace string, epServiceName string, hostips []string) (endpoint *api.Endpoints, service *api.Service, err error) {
+
+	addrlist := make([]api.EndpointAddress, len(hostips))
+	for i, v := range hostips {
+		addrlist[i].IP = v
+	}
+	endpoint = &api.Endpoints{
+		ObjectMeta: api.ObjectMeta{
+			Namespace: namespace,
+			Name:      epServiceName,
+		},
+		Subsets: []api.EndpointSubset{{
+			Addresses: addrlist,
+			Ports:     []api.EndpointPort{{Port: 1, Protocol: "TCP"}},
+		}},
+	}
+	_, err = p.plugin.host.GetKubeClient().Core().Endpoints(namespace).Create(endpoint)
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = nil
+	}
+	if err != nil {
+		glog.Errorf("glusterfs: failed to create endpoint %s", err)
+		return nil, nil, fmt.Errorf("error creating endpoint %v", err)
+	}
+	service = &api.Service{
+		ObjectMeta: api.ObjectMeta{Name: epServiceName, Namespace: namespace},
+		Spec: api.ServiceSpec{
+			Ports: []api.ServicePort{
+				{Protocol: "TCP", Port: 1}}}}
+	_, err = p.plugin.host.GetKubeClient().Core().Services(namespace).Create(service)
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = nil
+	}
+	if err != nil {
+		glog.Errorf("glusterfs: failed to create service %s", err)
+		return nil, nil, fmt.Errorf("error creating service %v", err)
+	}
+	return endpoint, service, nil
+}
+
+func (d *glusterfsVolumeDeleter) DeleteEndpointService(namespace string, epServiceName string) (err error) {
```

Comment:
Done. afaict, if we delete svc first , it will delete the ep as well. So I kept ep deletion followed by svc to make sure both are gone. Any way reverted the change and only keeping service delete.